### DISCUSSION TITLE
Sending multiple collections on one request

### DISCUSSION THREAD

**Discussion by theMackabu** [pocketbase/pocketbase#143] [Q&A] [answered]
Is it possible, for example, to send the "pages" and the "links" collection at the same time in a request? Is this possible via using pocketbase as a framework or does the feature exist already?

**Comment by ganigeorgiev** [comment, marked as answer]
Transactional create/update of multiple collections, records, etc. is not possible via the web API.

Bulk update/import for the collection schema and the records will be added in the future (see https://github.com/pocketbase/pocketbase/issues/48).

When you use PocketBase as a framework, you don't have to make http requests and can directly interact with the PocketBase internals and the database (this currently is not documented very well, but it is in the roadmap). For example, if you want to do multiple DB operations in a transactional manner (aka. save in the database only if all calls succeed), you could do something like this in your Go application:

```go
err := app.Dao().RunInTransaction(func(txDao *Dao) error {
    if err := txDao.SaveRecord(...); err != nil {
        return err
    }
    
    if err := txDao.SaveCollection(...); err != nil {
        return err
    }
    
    // ... any other db operation
    
    return nil
})
```

**Reply by theMackabu** [reply]
Thanks, but after looking into that, I had a small issue with `app.Dao()`, pocketbase [docs](https://pocketbase.io/docs/db-queries/) mention
```go
 app.Dao().FindCollectionByIdOrName("string here")
```
but I still seem to be getting this issue: ![Screen Shot 2022-07-16 at 3 54 00 PM](https://user-images.githubusercontent.com/32078755/179352836-d8c1fcff-27c3-4561-bcc9-e7ee64491d6e.png) and im not sure how to fix it.

**Reply by ganigeorgiev** [reply]
Oops, sorry. It should be `app.Dao().FindCollectionByNameOrId("...")`. I'll fix it in the documentation later today.

You could find all db methods by looking in the [/daos directory](https://github.com/pocketbase/pocketbase/tree/master/daos) or in the [godoc reference](https://pkg.go.dev/github.com/pocketbase/pocketbase#section-directories).

**Reply by theMackabu** [reply]
> FindCollectionByNameOrId

thanks

**Comment by theMackabu** [comment]
Sorry to ask again, but I cant seem to extend the api to send multiple collections on request (im new to go) 
I have tried about 20 different variations, all ending in errors of all sorts. If it isnt possible could you let me know?

**Reply by ganigeorgiev** [reply]
Could you clarify what do you mean by ` send multiple collections on request`.
Do you mean creating multiple records from a single collection?

As I said earlier, the web API (aka. the REST endpoins) doesn't support bulk operations at the moment.

You can do that only if you directly use the PocketBase database abstraction.

Could you share your code that doesn't work to understand better what actually the use case is?

**Reply by theMackabu** [reply]
> Could you clarify what do you mean by send multiple collections on request.

A request from the API would return multiple collections instead of just a single collection

some code I tried
```go
 	  	model := &models.Collection{}

		e.Router.GET("/api/classcircle/init", app.Dao().CollectionQuery().
		 AndWhere(dbx.Or(
			 dbx.HashExp{"name": "settings"},
		 )).Limit(1).One(model))
```
	
```go
 	  	model := &models.Collection{}

		e.Router.AddRoute(echo.Route{
			  Method: http.MethodGet,
			  Path:   "/api/classcircle/init",
			  Handler: func(c echo.Context) error {
					return app.Dao().CollectionQuery().
					 AndWhere(dbx.Or(
						 dbx.HashExp{"name": "settings"},
					 )).
					 Limit(1)
			  },
		 })
```

```go
		e.Router.AddRoute(echo.Route{
			  Method: http.MethodGet,
			  Path:   "/api/classcircle/init",
			  Handler: func(c echo.Context) error {
					return app.Dao().FindCollectionByIdOrName("settings") + app.Dao().FindCollectionByIdOrName("navigation")
			  },
		 })
```

**Reply by ganigeorgiev** [reply]
Action handlers must write something to the response body. Returning just the result of the dao does nothing if there is no error.

In addition, concatenating structures with `+` doesn't make sense since this is not a valid go code (I recommend installing a LSP extension in your editor to help you with autocomplete and other code suggestions).

Here is an example of your last code snippet that should work:

```go
package main

import (
	"log"
	"net/http"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
	// enable if you are using formatted errors
	// "github.com/pocketbase/pocketbase/rest"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodGet,
			Path:   "/api/classcircle/init",
			Handler: func(c echo.Context) error {
				collectionA, err := app.Dao().FindCollectionByNameOrId("settings")
				if err != nil {
					return err
					// or if you want a formattet error:
					// return rest.NewBadRequestError("Something went wrong", err);
				}

				collectionB, err := app.Dao().FindCollectionByNameOrId("navigation")
				if err != nil {
					return err
					// or if you want a formattet error:
					// return rest.NewBadRequestError("Something went wrong", err);
				}

				return c.JSON(http.StatusOK, map[string]any{
					"settings":   collectionA,
					"navigation": collectionB,
				})
			},
		})

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}

```

NB! The above may not return what actually you are trying to do. **Collections** are the structure of your data (eg. like table schema in sql). **Records** are actually the data of your application and probably that is what you want. **In addition, the above doesn't attach any middlewares so there would be no access checks and everyone will be able to access the returned data.**

Why do you want to use PocketBase as framework in the first place?
Does not simple client side code using the [js-sdk](https://github.com/pocketbase/js-sdk) not enough in your case?
Do you have any specific server-side logic that you want to implement and it's not achievable by just using the REST API?

In any case, if you really want to use PocketBase as a framework, I'd recommend first:
- go through the "A tour of Go" - https://go.dev/tour/welcome/1
- install an editor/plugin that will help you writing Go code
- read the basics of the PocketBase router, **echo** - https://echo.labstack.com/guide/routing/ (we are using v5 but it is mostly the same as in v4) to help you understand what the above example is doing
